### PR TITLE
Timeline::download_missing: fix handling of mismatched layer size

### PR DESCRIPTION
**Pull request by problame**
Before this patch, when we decide to rename a layer file to backup because of layer file size mismatch, we would not remove the layer from the layer map, but remote the on-disk file.

Because we  re-download the file immediately after, we simply end up with two layer objects in memory that reference the same file in the layer map.  So, GetPage() would work fine until one of the layers gets delete()'d. The other layer's delete() would then fail.

### CODE REVIEW

File: pageserver/src/tenant/timeline.rs [hunk 1]
```diff
@@ -1015,49 +1015,69 @@ impl Timeline {
...
 
-            let local_layer_path = timeline_dir.join(remote_layer_name.file_name());
+            // Is the local layer's size different from the size stored in the
+            // remote index file?
+            // If so, rename_to_backup those files & replace their local layer with
+            // a RemoteLayer so that we re-download them on-demand.
+            if let Some(local_layer) = local_layer {
+                let local_layer_path = local_layer.local_path();
+                assert!(local_layer_path.exists());
+                assert!(local_layer_path == timeline_dir.join(remote_layer_name.file_name()));
```
Comment (koivunej) [inline pageserver/src/tenant/timeline.rs:1033]:
```suggestion
                assert_eq!(local_layer_path, timeline_dir.join(remote_layer_name.file_name()));
```

File: pageserver/src/tenant/timeline.rs [hunk 2]
```diff
@@ -1015,49 +1015,69 @@ impl Timeline {
...
-                info!("remote layer {remote_layer_name:?} does not exist locally");
+                let local_layer_path = timeline_dir.join(remote_layer_name.file_name());
+                assert!(
+                    !local_layer_path.exists(),
+                    "caller did not supply exhaustive list of local_layers: {local_layer_path:?}"
+                );
+                // fall-through to adding the remote layer
             }
 
+            info!(
```
Comment (koivunej) [inline pageserver/src/tenant/timeline.rs:1076]: Assuming you will do the demotions from `info` => `debug` above, and we see only these log lines in the log, not the download progress, I think the outcome will be confusing.

I'd propose adding the next action to this message ("downloading"), and maybe an overall completion message with how many were downloaded before return?

**Review by koivunej — APPROVED** [review]
Overall this change looks good for me even though I might not remember the full context. Minor logging related suggestions.

**Review by SomeoneToIgnore — APPROVED** [review]
On the other end of this bug, there's `layer_map.rs::insert_historic` method that now does zero checks if duplicate layers are added: that may lead to whatever special effects, if we accidentally misuse this method.

We might want to fix it eventually too.

**Review by LizardWizzard — APPROVED** [review]
Looks good to me, plus one for @koivunej thoughts regarding info => debug.

**Comment by problame** [comment]
> On the other end of this bug, there's layer_map.rs::insert_historic method that now does zero checks if duplicate layers are added: that may lead to whatever special effects, if we accidentally misuse this method.
> We might want to fix it eventually too.

Are you referring to, before this PR, if `local_layer_path.exists()` we'd `already_downloaded = true; ...; continue`?

**Comment by SomeoneToIgnore** [comment]
I don't think this PR changes `insert_historic` behavior, but generally, anything, including the bug this PR ends up fixing, ends up bringing extra (duplicate) layers into layer map and we never notice that fast enough.

In your PR, you've mentioned

>  So, GetPage() would work fine until one of the layers gets delete()'d.

to me, we should fail fast the moment such layer is added to the layer map (or, at least log an error and discard the duplicate?). 
`insert_historic` is `pub fn` so it should expect whatever trash input may come, but it does not verify that at all.

But sure I don't require your PR to fix this bug, just a general note that we can make such errors more obvious and eager to appear in a relatively simple manner.

**Comment by problame** [comment]
> to me, we should never get in such state, rather failing fast at the moment such layer is added to the layer map (or, at least log an error and discard the duplicate?).

Agree.

**Pull request merged**